src/datagen/kidneyexchange.jl

```diff
@@ -0,0 +1,95 @@
+using StatsBase: sample
+
+struct KepGenerator <: AbstractModelGenerator 
+    nb_nodes::Int # probability of creating an edge between i and j
+    density::Real # number of pairs of the instance
+end
+
+"""
+    fill_with_generator!(cpmodel::CPModel, gen::KepGenerator; seed=nothing)   
+
+Fill a CPModel with the variables and constraints generated. We fill it directly instead of 
+creating temporary files for efficiency purpose!
+
+The decision matrix `x` is the adjacent matrix of the graph (i.e. the instance). 
+x[i, j] is branchable => pair i can receive a kidney from pair j
+x[i, j] is not branchable => pair i can not receive a kidney from pair j
+
+The generator ensure that all pairs have at least one incoming arc and one outgoing arc (i.e. at least one branchable variable per row and column).
+For this we first place one branchable variable in each line of the matrix with a different column index each time.
+Then, the remaining edges are assigned randomly between the non-asssigned elements of the decision matrix.
+"""
+function fill_with_generator!(model, gen::KepGenerator; seed=nothing)
+    if !isnothing(seed)
+        Random.seed!(seed)
+    end
+    density = gen.density
+    nb_nodes = gen.nb_nodes
+    total_edges = round(Int, nb_nodes*nb_nodes*density)
+    @assert total_edges > nb_nodes "density too low to ensure that all pairs have at least one incoming arc and one outgoing arc"
+
+    ### Instance ###
+
+    # ensure that all pairs have at least one in edge and one out edge 
+    indexes = [(i,j) for i = 1:nb_nodes for j = 1:nb_nodes]
+    permutation = shuffle([i for i=1:nb_nodes]) # permutation[i] = j => x[i, j] is branchable
+    flat_index_required_branchable = [permutation[i] + (i-1)*nb_nodes for i in 1:nb_nodes]
+    index_branchable = [] # Array that will containing the indexes (as tuples) of the branchable variables
+    for i in reverse(flat_index_required_branchable)
+        push!(index_branchable, splice!(indexes, i))
+    end
+
+    # the remaining edges are assigned randomly between the non-asssigned elements of the decision matrix
+    nb_unassigned_edges = total_edges - nb_nodes
+    append!(index_branchable, sample(indexes, nb_unassigned_edges))
```

Comment:
very elegant syntax! The fact that `splice!` removes the value at index `i` from the array ensures that you cannot append twice the same edge.

Reply:
Thanks :)